### PR TITLE
feat(cxx_indexer): Allow VNames as NodeIds for #5037

### PR DESCRIPTION
**Pull request by zrlk**
This PR allows VNames to be smuggled inside NodeIds. This is necessary
because the alternate semantic metadata rules need to be applied
inside IndexerASTHooks which only deals in NodeIds.

This implementation seemed slightly less evil than encoding a pointer
to a VName in a string. I considered using proto wire format, but
the lifetimes involved make things hard (the indexer likes to deal
in VNameRefs for efficiency (without them IIRC we spent a significant
percentage of cycles on copying strings around) and there's nowhere
for a temporary VName to go).

### CODE REVIEW

File: kythe/cxx/indexer/cxx/GraphObserver.h [hunk 1]
```diff
@@ -1059,6 +1062,42 @@ class GraphObserver {
...
+
+  VNameRef DecodeMintedVName(const NodeId& id) const {
+    const auto& bytes = id.getRawIdentity();
+    size_t path, root, corpus, language;
+    VNameRef ref;
+    CHECK(sizeof(size_t) * 4 <= bytes.size());
+    path = reinterpret_cast<const size_t*>(bytes.data())[0];
+    root = reinterpret_cast<const size_t*>(bytes.data())[1];
+    corpus = reinterpret_cast<const size_t*>(bytes.data())[2];
+    language = reinterpret_cast<const size_t*>(bytes.data())[3];
```
[the remark this answers is not in the document]
Comment (shahms) [reply kythe/cxx/indexer/cxx/GraphObserver.h:1089]: Or Coded{Input,Output}Stream from the proto io utilities (which is probably the better choice).

[the remark this answers is not in the document]
Comment (zrlk) [reply kythe/cxx/indexer/cxx/GraphObserver.h]: I suspected as much. Could you point me to where I can read more about this? I personally suspected something about alignment on unfriendly platforms but there's probably a much simpler proscription.

Is the latest push safe?

[the remark this answers is not in the document]
Comment (shahms) [reply kythe/cxx/indexer/cxx/GraphObserver.h]: It's a bit scattered throughout the standard, but the most direct reference is https://eel.is/c++draft/basic.lval#11

File: kythe/cxx/indexer/cxx/GraphObserver.cc [hunk 2]
```diff
@@ -43,4 +43,55 @@ std::string CompressString(absl::string_view InString, bool Force) {
...
+  std::memcpy(id.data(), &header, sizeof(MintedVNameHeader));
+  return NodeId::CreateUncompressed(getVNameClaimToken(), id);
+}
+
+VNameRef GraphObserver::DecodeMintedVName(const NodeId& id) const {
+  const auto& bytes = id.getRawIdentity();
+  VNameRef ref;
+  CHECK(sizeof(MintedVNameHeader) <= bytes.size());
+  MintedVNameHeader header;
+  memcpy(&header, bytes.data(), sizeof(MintedVNameHeader));
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/GraphObserver.cc:78]: `std::` (for consistency)

**Review by shahms — APPROVED** [review]

**Pull request merged**